src/mbgl/renderer/pattern_atlas.cpp

```diff
@@ -93,7 +93,7 @@ void PatternAtlas::upload(gfx::UploadPass& uploadPass) {
     //     return;
     // }
 
-    if (!atlasTexture2D) {
+    if (!atlasTexture2D || atlasTexture2D->getSize() != atlasImage.size) {
```

Comment:
Looking at this now, if we recreate the atlas after some number of drawables have already been created using it, they're going to have an old atlas texture. When shelfpack reruns, it might cause UV coords to shift enough to not match the older atlas instances in circulation. We should instead move the recreating to inside Texture2D and offer pointer stability here.